io_scene_niftools/modules/nif_import/armature/__init__.py

```diff
@@ -141,33 +142,29 @@ def store_bind_matrices(self, n_armature):
                         if isinstance(geom, NifClasses.BSTriShape):
                             vertex_data = geom.get_vertex_data()
                             if isinstance(geom, NifClasses.BSDynamicTriShape):
-                                # BSDynamicTriShape uses Vector4 to store vertices with a 0 W component, which would
-                                # nullify translation when multiplied by a Matrix44. Hence, first conversion to Vector3
-                                # and assign the position values back later.
-                                vertices = [NifClasses.Vector3.from_value((vertex.x, vertex.y, vertex.z)) for vertex in geom.vertices]
+                                vertices = geom.vertices
                             else:
                                 vertices = [vert_data.vertex for vert_data in vertex_data]
 
                             normals = [vert_data.normal for vert_data in vertex_data]
                         else:
                             vertices = geom.data.vertices
                             normals = geom.data.normals
-                        for vert in vertices:
-                            newvert = vert * diff
-                            vert.x = newvert.x
-                            vert.y = newvert.y
-                            vert.z = newvert.z
-                        diff33 = diff.get_matrix_33()
-                        for norm in normals:
-                            newnorm = norm * diff33
-                            norm.x = newnorm.x
-                            norm.y = newnorm.y
-                            norm.z = newnorm.z
-                        if isinstance(geom, NifClasses.BSDynamicTriShape):
-                            for vertex, t_vertex in zip(geom.vertices, vertices):
-                                vertex.x = t_vertex.x
-                                vertex.y = t_vertex.y
-                                vertex.z = t_vertex.z
+                        # BSDynamicTriShape uses Vector4 to store vertices with a 0 W component, which would
+                        # nullify translation when multiplied by a Matrix44. Hence, first conversion to three-component
+                        # vector
+                        np_vertices = np.array(vertices, dtype=float)[:,:3]
+                        np_vertices = np.concatenate((np_vertices, np.ones((len(np_vertices), 1), dtype=float)), axis=1)
```

Comment:
is the concat with 1 needed? you should be able to just multiply vec3 with the matrix...
and skip the [:, :3] below:
`for vertex, t_vertex in zip(vertices, np_vertices[:, :3]):`

Reply:
It is needed, otherwise you get an error. When you transform Vec3s with a Matrix like we did in the past, what happens in secret is that the matrix multiplication happens as though that 1 component is there. Blender probably still does that, because it's really only focused on matrix multiplication in the context of 3D transformations. Numpy, however, is more strict about that, and requires that the dimensions match.

If you try to make it match by reducing the matrix from a 4x4 matrix to a 3x3 matrix, you remove the translation component.

Reply:
Good catch, I was not aware of that. You're creating two arrays now (the ones and the final result), which you can bypass by using np.pad

https://numpy.org/doc/stable/reference/generated/numpy.pad.html#numpy.pad
probably:
np.pad(a, (0, 1), 'constant', constant_values=1.0)